### PR TITLE
Desktop Attributes

### PR DESCRIPTION
**Pull request by BryonLewis**
Adding in the capability to load/create and edit track attributes within the desktop version.
 TestData: https://viame.kitware.com/#/viewer/6006e01cf751cc24d135ec2b - has attributes on track #1 I believe.  Remember that on web version you won't see attributes unless they are added to the global list of attributes.  This isn't required for the desktop version.

Explanation:
- Process CSV import data to determine track attributes.  Currently done outside of the actual CSV importation which isn't the most performant but is organized a bit better.  I essentially took advantage of the outside loop used to process the `TrackData[]` into a `MultiTrackRecord`.  Let me know if you feel strongly about this being injected directly into the parser.  I like how the parser returns a clean Array right now.
- The AttributeProcessor is a function used to iterate through the `TrackData[]` and gather up the used attributes as well as their values and number of times those values occur.  It does a secondary pass of the attributes to try to determine the attribute type.  It cascades from `number` to `boolean` to `text` based on attempting to covert all of the possible set values for that attribute.  Finally it counts the number of occurrences of the value and if it is text and all values are used at least twice it will infer that it is a set of discrete values.
- When we have the ability to load the `meta.json` that will override any settings that are calculated with the csv/json file itself.
- Modified the API for `attributes` to include the `datasetId`.  This is necessary for desktop to access the proper metadata file to update it when attributes are edited.  Note these are typically one time things that occur when adding/editing/deleting an attribute from the system or on mount for the TrackDetailsView.  The web version ignores this data, but in the future will be used to directly modify the metadata attributes field utilizing the same endpoints instead of storing it globally in girder.

Questions:

- Should parsing for attributes be moved into CSV Parser or left with the `TrackData[]` -> `MultiTrackRecord` conversion?
- Should I remove the inferring of type and leave all items as text?
- Should the `attributeProcessor.ts` be it's own thing or moved into `utils.ts`?

TODO (Future PR):
- Setting/Updating/Deleting attributes both for the web version and the desktop version affect the displaying of attributes in the editor but not the source data of the track JSON.  Technically deleting an attribute should search all of the tracks and remove the attribute wherever it is set in the JSON field.  I'm wondering if that should be implemented.  This is an expensive task based on the number of detections and attributes in the system.  May want to have a discussion on how to implement this.

### CODE REVIEW

File: client/platform/desktop/frontend/components/Recent.vue [hunk 1]
```diff
@@ -30,6 +30,7 @@ export default defineComponent({
             });
           } else {
             // Display new data and await transcoding to complete
+            console.log(meta);
```
Comment (subdavis) [inline client/platform/desktop/frontend/components/Recent.vue:33]: remove

File: client/viame-web-common/components/Viewer.vue [hunk 2]
```diff
@@ -429,7 +429,7 @@ export default defineComponent({
       style="min-width: 700px;"
     >
       <sidebar
-        v-bind="{ newTrackSettings, typeSettings }"
+        v-bind="{ newTrackSettings, typeSettings, datasetId:id }"
```
Comment (subdavis) [inline client/viame-web-common/components/Viewer.vue:432]: It's possible that `provides.State` would have been a better place to put this, especially considering everything in state belongs to exactly one dataset.  I'm not going to dwell on this, it's good like it is for now.

File: client/platform/desktop/frontend/api.ts [hunk 3]
```diff
@@ -112,19 +112,25 @@ async function saveDetections(id: string, args: SaveDetectionsArgs) {
 /**
  * Unimplemented sections of the API
```
Comment (subdavis) [inline client/platform/desktop/frontend/api.ts:113]: This comment is incorrect now.

File: client/platform/desktop/frontend/api.ts [hunk 4]
```diff
@@ -112,19 +112,25 @@ async function saveDetections(id: string, args: SaveDetectionsArgs) {
...
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
-async function setAttribute({ addNew, data }: {addNew: boolean | undefined; data: Attribute}) {
-  return Promise.resolve();
+async function setAttribute(datasetId: string, { addNew, data }:
+  {addNew?: boolean; data: Attribute}) {
+  const client = await getClient();
+  return client.post(`dataset/${datasetId}/attribute`, { addNew, data });
 }
 
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (subdavis) [inline client/platform/desktop/frontend/api.ts:130]: these comments should be removed.

File: client/platform/desktop/backend/server.ts [hunk 5]
```diff
@@ -64,6 +64,41 @@ apirouter.post('/dataset/:id/meta', async (req, res, next) => {
   }
 });
 
+/* LOAD Attributes */
+apirouter.get('/dataset/:id/attribute', async (req, res, next) => {
```
Comment (subdavis) [inline client/platform/desktop/backend/server.ts:68]: Should probably note somewhere that this is not an ideal way of loading attributes.  By the time this is called, the attributes were loaded already during metadata load, so this loads metadata twice.

If we suppose that, once web is converted, attributes will always be part of the meta object, why not get rid of this method altogether?

File: client/platform/desktop/backend/native/common.ts [hunk 6]
```diff
@@ -321,10 +323,47 @@ async function saveMetadata(settings: Settings, datasetId: string, args: Dataset
...
     existing.customTypeStyling = args.customTypeStyling;
   }
+  if (args.attributes) {
+    existing.attributes = args.attributes;
+  }
   await _saveAsJson(projectDirInfo.metaFileAbsPath, existing);
   await release();
 }
 
+async function getAttributes(settings: Settings, datasetId: string):
```
Comment (subdavis) [inline client/platform/desktop/backend/native/common.ts:333]: Possibly remove?  See other comment.

**Review by subdavis — CHANGES_REQUESTED** [review]
Remove comments and console.log

Possibly remove `loadAttributes` and its API call.

**Comment by subdavis** [comment]
> Should parsing for attributes be moved into CSV Parser or left with the TrackData[] -> MultiTrackRecord conversion?

I like it how it is.

> Should I remove the inferring of type and leave all items as text?

Type inference seems handy, but if it turns out to be unreliable, we can remove it.

> Should the attributeProcessor.ts be it's own thing or moved into utils.ts?

I like it in its own file.

File: client/viame-web-common/components/Viewer.vue (hunk 2 above)
Comment (BryonLewis) [reply client/viame-web-common/components/Viewer.vue:432]: My brain was broken at the moment.  I knew internally that this is something that would be in a global store in vuex terms and the provides stuff just slipped my mind for some reason.  I'll look into modifying it and providing it from there.

File: client/platform/desktop/backend/server.ts (hunk 5 above)
Comment (BryonLewis) [reply client/platform/desktop/backend/server.ts:68]: It's this way because we aren't managing the attributes metadata directly in the loaded metadata of the webapp.  When we make a change is does a round trip to confirm that it was added properly to the metadata file.
Adding/Editing metadata calls a `setAttributes`, then a `getAttributes` right after to update the current list of attributes.  This was a holdover from utilizing the Global attributes in girder.  I could implement a `useAttributes` to modify the attributes local to the client and then await the response from setting to make sure it was done properly.  I think we may get there in the future and then implement the same upsert/delete sort of updating used for the tracks json.  I was trying to keep this somewhat simple to begin with.

I'll add a note to this that in the future this will probably be swapped for local management of metadata while hitting the canary for saving of the trackJSON and metadata at the same time when the user modifies attributes.

Comment (subdavis) [reply client/platform/desktop/backend/server.ts:68]: Makes sense to me, maybe add a comment, but no other change necessary.

**Comment by BryonLewis** [comment]
Did the comment updates and I created a `useDatasetId` to provide the datasetID to the track-details.  This is much better.

**Comment by subdavis** [comment]
Something weird is going on.

Track and detection attributes seem to erase each other.


https://user-images.githubusercontent.com/4214172/106312338-78d55380-6234-11eb-910c-8083ce070b56.mp4

**Comment by BryonLewis** [comment]
> Something weird is going on.
> 
> Track and detection attributes seem to erase each other.

Wasn't thinking right.  On girder the `_id` is provided for you when you create a new attribute hence you could get conflicts with the same name.  So it wasn't setting the `_id` properly.

**Review by subdavis — CHANGES_REQUESTED** [review]
Looks great.

One last thing: can we change the default visibility mode to always show the editable fields?  Its fine that you can have the nice condensed view, but if you create a new attribute, it doesn't show up until you hit the eye, which I don't think it a good default.

This optimization should be opt-in rather than opt-out.

**Comment by BryonLewis** [comment]
> Looks great.
> 
> One last thing: can we change the default visibility mode to always show the editable fields? Its fine that you can have the nice condensed view, but if you create a new attribute, it doesn't show up until you hit the eye, which I don't think it a good default.
> 
> This optimization should be opt-in rather than opt-out.

I agree that after adding a new attribute it should be immediately available.  I think we can change it so that if you add a new attribute it will swap to show all the fields after adding it.  The main concern I have is that there is a difference between viewing attributes and actively editing them.  Some of my test datasets have 10-15 track and then another 10-15 detection attributes.  They would all be displayed in the list by default meaning that it's difficult to view the data because of the sparseness especially if flipping through tracks or frames.  I'm open to any other alternatives if you think of any, maybe the eye should be something else?

image

**Comment by subdavis** [comment]
I'm just sort of looking at the risks:

"too busy, hard to parse, I'm overwhelmed" vs "I think this is broken it's not working I'm frustrated"

seems like we shoud choose the first risk. It's a hard problem and I can't come up with a solution that will work for everyone.

**Comment by subdavis** [comment]
There's always the "show more" / "show less" option at the end of the list.  Those are pretty ubiquitous, might be more intuitive than the eye, but I still think "show more" is a better default.
